### PR TITLE
Print error message if an invalid mapgen alias was detected

### PR DESCRIPTION
**Pull request by Wuzzy2**
This PR adds a couple of checks into the mapgen to see if any mapgen alias resolves to `CONTENT_IGNORE`.
If yes, an error message is reported.

## To do

This PR is Ready for Review.

## How to test

Create a game and intentionally put in mapgen aliases that are invalid.
Or just don't create any mapgen aliases as well (which is also incorrect most of the time).

Example:
```
minetest.register_alias("mapgen_stone", "a_node_that_does_not_exist")
minetest.register_alias("mapgen_cobble", "another_node_that_does_not_exist")
-- and so on ...
```

Test out different mapgens: v6, v5, v7, flat, … your choice.

## Downside
For some strange reason, when you first start Minetest in a game with invalid mapgen aliases, the error is properly logged but not screamed into the chat. You will see it on the 2nd start, however.

I don't know if I can do anything about it, my guess is that it's a limitation of the debug logging system maybe. But really I have no clue. It's a minor downside IMO, the fact that it's logged at all is still an improvement.

## Justification
Invalid mapgen aliases make the mapgen choke, and you will find `ignore` in the world where it shouldn't be. So it's critical to get the mapgen aliases right, as a gamedev.

Currently, if the gamedev forgets a mapgen alias, it might be tricky to figure out that this is the reason. It might also be very suble, if you only forgot `mapgen_apple` or `mapgen_cobble` in v6, you only occassionally run into `ignore` areas and in case of `mapgen_apple`, it's so subtle this bug can go unnoticed for a good amount of time, as `ignore` is completely invisible.

Reporting this as error should allow gamedevs to quickly eliminate this kind of bug.

### CODE REVIEW

**Comment by paramat** [comment]
Good idea, i will test and review after release.

File: src/mapgen/mapgen_v6.cpp [hunk 1]
```diff
@@ -132,6 +132,39 @@ MapgenV6::MapgenV6(MapgenV6Params *params, EmergeManager *emerge)
 		c_stair_cobble = c_cobble;
 	if (c_stair_desert_stone == CONTENT_IGNORE)
 		c_stair_desert_stone = c_desert_stone;
+
+	if (c_stone == CONTENT_IGNORE)
+		errorstream << "Mapgen v6: Mapgen alias 'mapgen_stone' is invalid!" << std::endl;
```
Comment (HybridDog) [inline src/mapgen/mapgen_v6.cpp:137]: This is a lot of duplicate code, and there exists an additional `getId` method which returns a success boolean:
http://github.com/minetest/minetest/blob/3d6b55d3e9b91e6afdbfdcea0268dfb88c69c1c2/src/nodedef.h#L493-L507
http://github.com/minetest/minetest/blob/3d6b55d3e9b91e6afdbfdcea0268dfb88c69c1c2/src/nodedef.cpp#L1007-L1023

Maybe it is possible to use [lambda expressions](https://en.cppreference.com/w/cpp/language/lambda) to shorten the code, and set the node to air if the mapgen alias is missing.
```c++
[…]
	auto getIdValid = [&](const std::string &name)->content_t
	{
		content_t id = CONTENT_AIR;
		if (!ndef->getId(name, id))
			errorstream << "Mapgen v6: Mapgen alias '" << name <<
				"' is invalid!" << std::endl;
		return id;
	}

	c_stone           = getIdValid("mapgen_stone");
	c_dirt            = getIdValid("mapgen_dirt");
[…]
```

File: src/mapgen/mapgen.cpp [hunk 2]
```diff
@@ -629,6 +629,17 @@ MapgenBasic::MapgenBasic(int mapgenid, MapgenParams *params, EmergeManager *emer
...
 		c_lava_source = c_water_source;
+
+	if (c_stone == CONTENT_IGNORE)
+		errorstream << "Mapgen: Mapgen alias 'mapgen_stone' is invalid!" << std::endl;
+	if (c_water_source == CONTENT_IGNORE)
+		errorstream << "Mapgen: Mapgen alias 'mapgen_water_source' is invalid!" << std::endl;
+	if (c_river_water_source == CONTENT_IGNORE)
+		warningstream << "Mapgen: Mapgen alias 'mapgen_river_water_source' is invalid!" << std::endl;
+	if (c_lava_source == CONTENT_IGNORE)
+		warningstream << "Mapgen: Mapgen alias 'mapgen_lava_source' is invalid!" << std::endl;
```
Comment (paramat) [inline src/mapgen/mapgen.cpp:640]: This and the one above are 'warningstream' instead of 'errorstream'.

**Comment by Wuzzy2** [comment]
I think lambda expressions violate the “Don't be too C++y” coding style rule.

Please, when reviewing this, also make sure whether the specific errors make sense. I mean, Minetest should obviously not report errors for optional mapgen aliases (e.g. when there is a fallback).
I'm not sure if some of the aliases I report errors for are actually optional (like lava and river water).

**Comment by HybridDog** [comment]
> I think lambda expressions violate the “Don't be too C++y” coding style rule.

In this case, a new method in NodeDefManager could fit better than lambda expressions because the function is required in several mapgen code files.

**Comment by paramat** [comment]
> Minetest should obviously not report errors for optional mapgen aliases (e.g. when there is a fallback).
I'm not sure if some of the aliases I report errors for are actually optional (like lava and river water).

Aha thanks for reminding me, which are essential is detailed here:
https://github.com/minetest/minetest/blob/45a2ca0af2285ab32003bd0d07690551b895b480/doc/lua_api.txt#L293-L343

So in non-mgv6 mapgens, only 3 are essential so only those 3 should trigger an error.
In mgv6, all are essential.

**Comment by sfan5** [comment]
> I think lambda expressions violate the “Don't be too C++y” coding style rule.

I disagree, lambdas are a great way to deduplicate code inside a method and despite being a recent C++ feature aren't overly complex.
I mostly associate that rule with unnecessary template use.

Also if you grep the source for `=\s*\[.*\]\s*\(.*\)`, you will notice that lambda expression are already used in numerous places.

**Comment by paramat** [comment]
Just realised something ...
> and set the node to air if the mapgen alias is missing, or add a "get valid id or air" method to nodedef.

Please do not do this, it will cause bugs.

Throughout mapgen code (and in the lines above), a fundamental mapgen node being CONTENT_IGNORE is a flag for an undefined mapgen alias, critical mapgen decisions are made based on this.

It could then be suggested that these decisions are made by checking for CONTENT_AIR instead, but no, because "air" is a valid node for a mapgen alias, so mapgen code then has no idea whether the fundamental mapgen node is a valid defined mapgen alias or an undefined mapgen alias.
CONTENT_IGNORE is the only content ID (other than CONTENT_UNKNOWN) where you can be sure it is an undefined mapgen alias, as "ignore" is always an invalid fundamental mapgen node.

Also for consistency, as we will only be checking 3 fundamental mapgen nodes, not all 5, so the content ID flag for an undefined mapgen alias would then be different for certain nodes, very messy and confusing.

**Comment by Wuzzy2** [comment]
> In mgv6 (which uses treegen.cpp), all the listed ones are essential.

Are you sure? Based on my tests, I can make MGv6 run perfectly with only these aliases:

> mapgen_stone
> mapgen_water_source
> mapgen_lava_source
> mapgen_dirt
> mapgen_dirt_with_grass
> mapgen_sand
> mapgen_tree
> mapgen_leaves
> mapgen_apple
> mapgen_cobble

And reading the code confirms this. For other aliases like `mapgen_ice` or `mapgen_pine_tree` there's a fallback, so if I don't define an alias in Lua, the v6 mapgen is still functional. Also, current `minimal` also doesn't set all the v6 mapgen aliases.

Anyway. As for the PR, I think I will work on it a while later. There's still a release waiting … Or do you want this to be done before release? Just that I understand correctly: Your only complaint right now is that the cose is kinda verbose/redundant? Or did I miss something?

**Comment by paramat** [comment]
> Are you sure?

Actually, you are correct, some mgv6 aliases are optional. When i wrote the docs i probably could not be bothered to divide mgv6 aliases in to 'essential' and 'optional'.
Also there are other factors, for example if snowbiomes are not enabled, all pine, snow and ice aliases are unnecessary.
If biome noises are customised some biomes may never occur, which also changes the essential alises.
So it is difficult to decide which aliases to check.

Strictly, in non-mgv6 mapgens, if Mapgen Valleys is not used, the 'river water' alias is not essential. But i prefer that is still checked as it will be used by Mapgen Watershed also if that is merged (it is nearly ready) and by future mapgens which hopefully will have sloping rivers.

Unfortunately the result of all the usage options means that only a few aliases can be checked.
Also for some users, error messages will be triggered for aliases the user actually does not need.

**Comment by paramat** [comment]
> Just that I understand correctly: Your only complaint right now is that the cose is kinda verbose/redundant? Or did I miss something?

Mgv6:
The other core devs may prefer making the code in mgv6 more compact using a lambda expression.
Personally i do not mind how you have coded this, as far fewer aliases actually need to be checked, so your code will become less verbose.

Non-mgv6:
I can see a benefit of a lambda function for mgv6, but for non-mgv6 mapgens only 3 aliases need checking, so i prefer the current simple conditionals for non-mgv6 mapgens, more readable and fewer lines.

File: src/mapgen/mapgen_v6.cpp (hunk 1 above)
Comment (paramat) [reply src/mapgen/mapgen_v6.cpp:137]: I am neutral about using a lambda expression, but it should only be used for mgv6.
Non-mgv6 mapgens only need 3 aliases checked, so in that case simple repeated conditionals will be fewer lines and easier to understand.

Also, why does the lambda expression example above set content ID to CONTENT_IGNORE? If the alias is missing the content ID of the mapgen node will already be CONTENT_IGNORE.
Please only add code to trigger an error, nothing else.

To be honest i am still confused by HybridDog's code example.
I am not sure `ndef->getId(name, id)` is needed because `ndef->getId(name)` looks just like the code example above, see
https://github.com/minetest/minetest/blob/3d6b55d3e9b91e6afdbfdcea0268dfb88c69c1c2/src/nodedef.cpp#L1018-L1023

Comment (HybridDog) [reply src/mapgen/mapgen_v6.cpp:137]: > If the alias is missing the content ID of the mapgen node will already be CONTENT_IGNORE.

No, the getId method with two arguments does not set it to ignore automatically.

> I am not sure `ndef->getId(name, id)` is needed […]

The getId method with one argument is just a wrapper for the other one. 
Instead of adding a lambda expression, an additional wrapper could be added to NodeDefManager, which can be used in all the mapgen files.

Comment (paramat) [reply src/mapgen/mapgen_v6.cpp:137]: But lines 95-107 have already set the ID to ignore if the alias is missing ... ?

Is your code example meant to be inserted at line 136 or line 95? I assumed 136.
If you mean line 95 the code is wrong, because checking for a missing alias must be done after the fallbacks are applied, not before.
For example if the stone alias is present but the desert_stone alias is missing, your code will trigger an error even though afterwards c_desert_stone falls back to c_stone, which is fine and should not trigger an error.

If you mean line 136 then the code is also wrong for reasons already explained.

I will close this comment for now but feel free to rethink, fix and resubmit a code suggestion.

**Comment by paramat** [comment]
Wuzzy2, update on the last comment:

HybridDog's code suggestion is wrong, please do not use it.
Here are my requests:

Mgv6:
I do not mind the use of a lambda expression (if necessary) or function to shorten your code, but:
It should be placed at line 136 after the existing code, it should not replace the existing code.
It should simply be used to make your code more compact, and do nothing else.

Non-mgv6:
As only 3 mapgen nodes need checking, your current code is fine as it is simpler, easier to understand and will actually have less lines.
Because the river water alias is semi-optional that one could perhaps trigger a warning instead of an error.

Treegen:
Looks fine.

File: src/mapgen/mapgen.cpp [hunk 3]
```diff
@@ -629,6 +629,17 @@ MapgenBasic::MapgenBasic(int mapgenid, MapgenParams *params, EmergeManager *emer
 	// Lava falls back to water as both are suitable as cave liquids.
 	if (c_lava_source == CONTENT_IGNORE)
 		c_lava_source = c_water_source;
+
+	if (c_stone == CONTENT_IGNORE)
+		errorstream << "Mapgen: Mapgen alias 'mapgen_stone' is invalid!" << std::endl;
+	if (c_water_source == CONTENT_IGNORE)
+		errorstream << "Mapgen: Mapgen alias 'mapgen_water_source' is invalid!" << std::endl;
+	if (c_river_water_source == CONTENT_IGNORE)
+		errorstream << "Mapgen: Mapgen alias 'mapgen_river_water_source' is invalid!" << std::endl;
```
Comment (paramat) [inline src/mapgen/mapgen.cpp:638]: As river water is semi-optional (only 1 mapgen requires it) this could be changed to a warning.

File: src/mapgen/mapgen.cpp [hunk 4]
```diff
@@ -629,6 +629,17 @@ MapgenBasic::MapgenBasic(int mapgenid, MapgenParams *params, EmergeManager *emer
...
+	if (c_lava_source == CONTENT_IGNORE)
+		errorstream << "Mapgen: Mapgen alias 'mapgen_lava_source' is invalid!" << std::endl;
+	if (c_cobble == CONTENT_IGNORE)
+		errorstream << "Mapgen: Mapgen alias 'mapgen_cobble' is invalid!" << std::endl;
```
Comment (paramat) [inline src/mapgen/mapgen.cpp:642]: These checks for lava source and cobble should be removed, as cave liquids and dungeon nodes are now usually defined in biome defintions, not by mapgen aliases.

**Comment by paramat** [comment]
In mgv6, due to fallbacks, only these mapgen nodes need checking:

stone
dirt
dirt with grass
sand
water source
lava source
cobble

As there are only 7, your current code is probably no longer than using a lambda expression or function, so i prefer your current way of coding this as it is simpler and easier to understand.

File: src/mapgen/mapgen_v6.cpp [hunk 5]
```diff
@@ -132,6 +132,39 @@ MapgenV6::MapgenV6(MapgenV6Params *params, EmergeManager *emerge)
...
+	if (c_ice == CONTENT_IGNORE)
+		errorstream << "Mapgen v6: Mapgen alias 'mapgen_ice' is invalid!" << std::endl;
+	if (c_cobble == CONTENT_IGNORE)
+		errorstream << "Mapgen v6: Mapgen alias 'mapgen_cobble' is invalid!" << std::endl;
+	if (c_mossycobble == CONTENT_IGNORE)
+		errorstream << "Mapgen v6: Mapgen alias 'mapgen_mossycobble' is invalid!" << std::endl;
+	if (c_stair_cobble == CONTENT_IGNORE)
+		errorstream << "Mapgen v6: Mapgen alias 'mapgen_stair_cobble' is invalid!" << std::endl;
+	if (c_stair_desert_stone == CONTENT_IGNORE)
+		errorstream << "Mapgen v6: Mapgen alias 'mapgen_stair_desert_stone' is invalid!" << std::endl;
```
Comment (paramat) [inline src/mapgen/mapgen_v6.cpp:167]: Due to fallbacks, only these mapgen nodes need checking:

stone
dirt
dirt with grass
sand
water source
lava source
cobble

File: src/mapgen/mapgen_v6.cpp (hunk 1 above)
Comment (HybridDog) [reply src/mapgen/mapgen_v6.cpp:137]: > If you mean line 95 the code is wrong […]

I thought about inserting it to line 95 and using it only for the IDs which don't have an alias.

**Comment by Wuzzy2** [comment]
I have applied all of paramat's suggestions.

I have removed the river water check completely, as it turns out, river water actually falls back to air.

File: src/mapgen/mapgen.cpp (hunk 3 above)
Comment (paramat) [reply src/mapgen/mapgen.cpp]: Or the check removed, either is fine. Resolved.

File: src/mapgen/mapgen.cpp (hunk 4 above)
Comment (paramat) [reply src/mapgen/mapgen.cpp]: Done.

File: src/mapgen/mapgen_v6.cpp (hunk 5 above)
Comment (paramat) [reply src/mapgen/mapgen_v6.cpp]: Done.

**Comment by paramat** [comment]
Thanks, this looks fine now, will try to test and approve soon.

**Comment by paramat** [comment]
> as it turns out, river water actually falls back to air.

Where is the code that does that? I cannot find it and am sceptical of this.
A check might be needed after all, although of course it is not high priority as few mapgens use river water.

**Comment by Wuzzy2** [comment]
```
getIdFromNrBacklog(&c_river_water,   "mapgen_river_water_source", CONTENT_AIR,    false);
```

`CONTENT_AIR` is the fallback argument.

**Comment by paramat** [comment]
Ok you mean there.
That is not actually a fallback for mapgen nodes, it is a fallback for biome nodes, such as 'top', 'filler', 'riverbed', 'dust' etc. The biome-generating code is independent of base mapgen code and runs after base mapgen.
For example, if you disable biome generation in mg_flags, mgvalleys will create a world consisting only of air, stone, water and river water. The river water mapgen alias is one of the fundamental base mapgen nodes, so it should have a check in this PR.

**Comment by Wuzzy2** [comment]
Done. As a warning, as it was previously requested to be a warning.

File: src/mapgen/treegen.cpp [hunk 6]
```diff
@@ -44,6 +44,12 @@ void make_tree(MMVManip &vmanip, v3s16 p0, bool is_apple_tree,
 	MapNode treenode(ndef->getId("mapgen_tree"));
 	MapNode leavesnode(ndef->getId("mapgen_leaves"));
 	MapNode applenode(ndef->getId("mapgen_apple"));
+	if (treenode == CONTENT_IGNORE)
```
Comment (sfan5) [inline src/mapgen/treegen.cpp:47]: does this comparison even work?
treenode is a `MapNode` (struct) and CONTENT_IGNORE is a number

Comment (SmallJoker) [reply src/mapgen/treegen.cpp:47]: Apparently it does, when `CONTENT_IGNORE` is constructed as MapNode and then compared to `treenode`. Strange that it seems to happen automatically.

https://github.com/minetest/minetest/blob/12275e235a17e84805f4b3b667e1ddbdf9499fc6/src/mapnode.h#L142-L153

**Comment by paramat** [comment]
I did some experimentation in MT master to see how it behaves with missing aliases.

For non-mgv6 mapgens with missing aliases:

If biome generation is disabled, the initial mapgen places "ignore" for the missing aliases.

If biome generation is enabled, the initial mapgen places "ignore" for the missing aliases, then biome generation replaces those ignore nodes with biome nodes, which often fallback to "air", according to:
https://github.com/minetest/minetest/blob/12275e235a17e84805f4b3b667e1ddbdf9499fc6/src/mapgen/mg_biome.cpp#L324-L338
Which is why river water becomes "air" in Mapgen Valleys. Similarly, if the stone alias is missing, stone becomes "air".
However, because various missing aliases all become "ignore", this situation can be very messy and should trigger errors or warnings (as the PR now does).
In other words, checks are needed to avoid "ignore" being placed in the initial mapgen, this can be tested by disabling biome generation in 'mg_flags'.

**Comment by paramat** [comment]
For mgv6 with missing aliases:
Mapgen places "ignore" for the missing aliases.

**Comment by paramat** [comment]
:+1: Tested. Code is fine.
